dev/com.ibm.ws.security.acme/src/com/ibm/ws/security/acme/internal/AcmeProviderImpl.java

```diff
@@ -881,5 +966,71 @@ protected void updateAcmeConfigService(AcmeConfigService acmeConfigService, Map<
 	public void setAcmeApplicationStateListener(AcmeApplicationStateListener acmeApplicationStateListener) {
 		applicationStateListenerRef.set(acmeApplicationStateListener);
 	}
+	
+	/**
+	 * Set the Scheduler service ref
+	 */
+	@Reference(name = "scheduledExecutorService", service = ScheduledExecutorService.class, target = "(deferrable=false)")
+	protected void setScheduledExecutorService(ServiceReference<ScheduledExecutorService> ref) {
+		scheduledExecutorServiceRef.setReference(ref);
+	}
+
+	/**
+	 * Unset the scheduler ref and stop the certificate checker
+	 */
+	protected void unsetScheduledExecutorService(ServiceReference<ScheduledExecutorService> ref) {
+		if (acmeCertChecker != null) {
+			acmeCertChecker.stop();
+		}
+		scheduledExecutorServiceRef.unsetReference(ref);
+	}
+
+	/**
+	 * Get the scheduler ref
+	 */
+	public ScheduledExecutorService getScheduledExecutorService() {
+		return scheduledExecutorServiceRef.getService();
+	}
 
+    /**
+     * Acquire the writer lock. To be used to prevent concurrent certificate
+     * renew or revoke requests. Must be used with releaseWriteLock
+     */
+    @Trivial
+    void acquireWriteLock() {
+		rwRenewCertLock.writeLock().lock();
+    }
+    
+    /**
+     * Release the writer lock. To be used to prevent concurrent certificate
+     * renew or revoke requests. Must be used with acquireWriteLock
+     */
+    @Trivial
+    void releaseWriteLock() {
+    	rwRenewCertLock.writeLock().unlock();
+    }
+    
+    /**
+     * Checks whether certificate renewal is allowed. It is allowed if:
+     * <li>Certificate renewal checking is disabled</li>
+     * <li>This is the first certificate request</li>
+     * <li>Enough time has passed since the last renewal</li>
+     * 
+     * If certificate renewal is not allowed, an exception is thrown.
+     */
+    @Override
+    public void checkCertificateRenewAllowed() throws CertificateRenewRequestBlockedException {
+    	long timeDiff = System.currentTimeMillis() - lastCertificateRenewalTimestamp;
+    	if (acmeConfig.isDisableMinRenewWindow() || lastCertificateRenewalTimestamp == -1 || (timeDiff >= AcmeConstants.RENEW_CERT_MIN)) {
+    		return;
+    	}
+
+    	if (tc.isDebugEnabled()) {
+    		Tr.debug(tc, "Too soon to renew, last certificate renewal was " + lastCertificateRenewalTimestamp);
+    	}
+    	CertificateRenewRequestBlockedException cr = new CertificateRenewRequestBlockedException("Too soon to renew, last certificate renewal was " + lastCertificateRenewalTimestamp);
```

Comment:
I would maybe make the constructor for CRRBE take the time remaining so it is always forced to be set. Doubt anywhere else will use this exception though.

Reply:
Added